### DISCUSSION TITLE
Download Only Providers

### DISCUSSION THREAD

**Discussion by Mimetis** [Mimetis/Dotmim.Sync#642] [Ideas]
As a sample, I've created two new providers, to demonstrate how to create your own provider, based on the ones already available.
These 2 providers are `SqliteSyncDownloadOnlyProvider` and `SqlSyncDownloadOnlyProvider`.

You can see the implementation of these 2 providers here : [Custom Providers sample](https://github.com/Mimetis/Dotmim.Sync/tree/master/Samples/CustomProvider)

> **Warning:** Making these providers required some minor changes in `DMS` core & existing providers. So far if you want to test them, please do so from the master branch source code.

``` cs
/// <summary>
/// Use this provider if your client database does not need to upload any data to the server.
/// This provider does not create any triggers / tracking tables and only 3 stored proc / tables.
/// If your client database is not readonly, any changes from server will overwrite client changes.
/// </summary>
public class SqlSyncDownloadOnlyProvider : SqlSyncProvider
{
    public SqlSyncDownloadOnlyProvider() { }
    public SqlSyncDownloadOnlyProvider(string connectionString) : base(connectionString) { }
    public SqlSyncDownloadOnlyProvider(SqlConnectionStringBuilder builder) : base(builder) { }

    .... 
    Code 
    Code 
    More code ....
    ....
}
```

Basically, they are:
- Only intended to be used on the client side (obvious for sqlite, but not for sql server).
- Will only download data, and will never upload anything.
- Really useful if you need a readonly client database.
- All changes on the client will be overwritten with server changes without any conflicts resolution (server wins.. always).
- No need to specify `SyncDirection=UploadOnly` on each table (it's by design).

Since client will never upload data, we are less dependent to tracking rows.
So far:
- Client triggers are not created.
- Tracking tables are not created.
- For Sql Server, only 3 stored procedures are created (per table).
- `scope_info` table is still needed and is still created.
- Update & Delete commands are way faster (since there is no more comparison with tracking rows).

![image](https://user-images.githubusercontent.com/4592555/143469520-d278820d-3478-4b05-8904-79b75f931059.png)

Less infrastructure, less complex commands means: better performances.

I've made 2 tests. Server is already provisioned. So far, the total duration is only due to client workload.
Here is 2 tests:
- First one using `SqlSyncDownloadOnlyProvider`
- Second one using `SqlSyncProvider`

``` bash
Synchronization done.
        Total changes  uploaded: 0
        Total changes  downloaded: 2752
        Total changes  applied: 2752
        Total resolved conflicts: 0
        Total duration :00.00:00.573
```
``` bash
Synchronization done.
        Total changes  uploaded: 0
        Total changes  downloaded: 2752
        Total changes  applied: 2752
        Total resolved conflicts: 0
        Total duration :00.00:00.939
```
Question: Is it something interesting for you guys ?

**Comment by workgroupengineering** [comment]
I think it is useful in a scenario where a catalogue is to be made

**Comment by gaurakshay** [comment]
@Mimetis for us, we need full syncing capabilities because some business logic that can change data lies on the mobile app and some lives on the server. But this very neat and certainly can be useful in some scenarios that we haven't thought of yet :)
